### PR TITLE
Network improvements

### PR DESCRIPTION
**Pull request by teclator**
## Description

Network was one of the first components we added to Agama and since then it the components have evolve a lot being one of the most out of date ones. Therefore it was known it deserve some love... and we tried to start revamping it a little bit.

**Changes done:**

  - Modified the Network Page using routes instead of popup modals to configure the connections or to connecto a WiFi network.
  - Modified the wifi selection list and added a drawer for showing the selected network description and connection options.
  - Notify device state reason changes in order to detect an authentication error when connecting to a WiFi (reopen the form is still pending)
  - Fixed gateway4 set up.
  - Do not show externally configured connections (libvirt, vpns..)

Manually tested

**Pending:**

  - Try to improve the state management as network changes usually close de opened websocket opening the door to miss state changes notifications. Because of that, the read of the devices, connections and WiFi networks is done  everytime insted of modifying the current state with just the changes.
  - Show the form of the failing authentication WiFi.
  - Fix the unit tests.
  - Remove unnecesary files, and fix types descriptions.

### Old UI

Screenshot from 2024-06-26 10-41-09
Screenshot from 2024-06-26 10-41-12
Screenshot from 2024-06-26 10-41-21
Screenshot from 2024-06-26 10-41-33

### New UI

Screenshot from 2024-06-26 10-49-08
Screenshot from 2024-06-26 10-49-19
Screenshot from 2024-06-26 10-49-22
Screenshot from 2024-06-26 10-49-34
Screenshot from 2024-06-26 10-49-40

### CODE REVIEW

File: web/src/components/network/NetworkPage.jsx [hunk 1]
```diff
@@ -58,135 +43,117 @@ const NoWiredConnections = () => {
...
  */
 export default function NetworkPage() {
   const { network: client } = useInstallerClient();
-  const { connections: initialConnections, settings } = useLoaderData();
+  const { connections: initialConnections, devices: initialDevices, settings } = useLoaderData();
   const [connections, setConnections] = useState(initialConnections);
-  const [devices, setDevices] = useState(undefined);
-  const [selectedConnection, setSelectedConnection] = useState(null);
+  const [devices, setDevices] = useState(initialDevices);
+  const [updateState, setUpdateState] = useState(false);
```
Comment (imobachgs) [inline web/src/components/network/NetworkPage.jsx:55]: Is `updateState` used? Perhaps a better name could help to know the intention.

File: web/src/components/network/NetworkPage.jsx [hunk 2]
```diff
@@ -58,135 +43,117 @@ const NoWiredConnections = () => {
...
+    const { wireless_enabled: wifiAvailable } = settings;
 
-    if (wifiConnections.length === 0) {
+    if (!wifiAvailable) {
       return (
-        <NoWifiConnections wifiScanSupported={settings.wireless_enabled} />
+        <CardField>
+          <CardField.Content>
+            <EmptyState title={_("No WiFi support")} icon="wifi_off" color="warning-color-200">
+              {_("The system does not support WiFi connections, probably because of missing or disabled hardware.")}
```
Comment (imobachgs) [inline web/src/components/network/NetworkPage.jsx:95]: Put in the shoes of an inexperienced user... I do not know if "missing hardware" is good enough.

File: web/src/components/network/NetworkPage.jsx [hunk 3]
```diff
@@ -58,135 +43,117 @@ const NoWiredConnections = () => {
...
+        <CardField.Content>
+          {activeConnection
+            ? (
+              <EmptyState title={sprintf(_("Conected to %s"), activeConnection.id)} icon="wifi" color="success-color-100">
+                {connectionAddresses(activeConnection)}
+              </EmptyState>
+            )
+            : (
+              <EmptyState title={_("No connected yet")} icon="wifi_off" color="color-300">
+                {_("The system has not been configured for connecting to a WiFi network yet.")}
```
Comment (imobachgs) [inline web/src/components/network/NetworkPage.jsx:124]: Please, use `Wi-Fi` (refer to [SUSE Documentation Style Guide](https://documentation.suse.com/style/current/single-html/docu_styleguide/index.html)).

File: web/src/components/network/WifiNetworkListItemPage.jsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) [2022] SUSE LLC
```
Comment (imobachgs) [inline web/src/components/network/WifiNetworkListItemPage.jsx]:
```suggestion
 * Copyright (c) [2024] SUSE LLC
```

Comment (imobachgs) [reply web/src/components/network/WifiNetworkListItemPage.jsx]: Or `2022-2024` if that's the case.

File: web/src/components/network/NetworkPage.jsx [hunk 5]
```diff
@@ -58,135 +43,117 @@ const NoWiredConnections = () => {
...
-    const wifiConnections = connections.filter(c => c.wireless);
+    const { wireless_enabled: wifiAvailable } = settings;
 
-    if (wifiConnections.length === 0) {
+    if (!wifiAvailable) {
       return (
-        <NoWifiConnections wifiScanSupported={settings.wireless_enabled} />
+        <CardField>
+          <CardField.Content>
+            <EmptyState title={_("No WiFi support")} icon="wifi_off" color="warning-color-200">
```
Comment (dgdavid) [inline web/src/components/network/NetworkPage.jsx:94]:
```suggestion
            <EmptyState title={_("No WiFi support")} icon="error">
```

File: web/src/components/network/WifiNetworkListItemPage.jsx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      // TRANSLATORS: Wifi network status
+      return _("Disconnected");
+    default:
+      return "";
+  }
+};
+
+const isStateChanging = (network) => {
+  const state = network.device?.state;
+  return state === DeviceState.CONFIG || state === DeviceState.NEEDAUTH || state === DeviceState.DEACTIVATING || state === DeviceState.FAILED;
```
Comment (imobachgs) [inline web/src/components/network/WifiNetworkListItemPage.jsx]: np: You used an `Array.includes` in a similar situation before.

File: web/src/components/network/WifiNetworkListItemPage.jsx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+};
+
+/**
+ * Component for displaying a Wi-Fi network within a NetworkList
+ *
+ * @param {object} props - component props
+ * @param {object} props.networks - the ap/configured network to be displayed
```
Comment (imobachgs) [inline web/src/components/network/WifiNetworkListItemPage.jsx]: `networks` or `network`?

File: web/src/components/network/WifiNetworkListItemPage.jsx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+ * @param {boolean} [props.isSelected] - whether the network has been selected by the user
+ * @param {boolean} [props.isActive] - whether the network is currently active
+ * @param {function} props.needAuth - function to execute when the network requires authentication
+ * @param {function} props.onSelect - function to execute when the network is selected
+ * @param {function} props.onCancel - function to execute when the selection is cancelled
+ */
+function WifiNetworkListItemPage({ network, isSelected, isActive, onSelect, onCancel }) {
+  // Do not wait until receive the next D-Bus network event to have the connection object available
```
Comment (imobachgs) [inline web/src/components/network/WifiNetworkListItemPage.jsx]: What about *until receiving*? And about D-Bus... well, from the client's POV is not D-Bus anymore ;-)

File: web/src/components/network/WifiNetworkListItemPage.jsx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+};
+
+/**
+ * Component for displaying a Wi-Fi network within a NetworkList
+ *
+ * @param {object} props - component props
+ * @param {object} props.networks - the ap/configured network to be displayed
+ * @param {boolean} [props.isSelected] - whether the network has been selected by the user
+ * @param {boolean} [props.isActive] - whether the network is currently active
+ * @param {function} props.needAuth - function to execute when the network requires authentication
```
Comment (imobachgs) [inline web/src/components/network/WifiNetworkListItemPage.jsx]: Is it still used?

File: web/src/components/network/WifiNetworkListItemPage.jsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+              <Icon name="lock" size="10" fill="grey" /> {network.security.join(", ")}{" "}
+              <Icon name="signal_cellular_alt" size="10" fill="grey" /> {network.strength}
+            </>
+          }
+          isChecked={isSelected || isActive || false}
+          onClick={onSelect}
+        />
+        <Split hasGutter>
+          {/* TRANSLATORS: %s is replaced by a WiFi network name */}
+          {false && showSpinner && <Spinner size="md" aria-label={sprintf(_("%s connection is waiting for an state change"), network.ssid)} />}
```
Comment (imobachgs) [inline web/src/components/network/WifiNetworkListItemPage.jsx]: `false`?

Comment (imobachgs) [reply web/src/components/network/WifiNetworkListItemPage.jsx]: What does "waiting for a state change" mean?

File: web/src/components/network/WifiNetworkListItemPage.jsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+          {showSpinner && network.device && network.device?.state === DeviceState.CONFIG}
+          <Text component="small" className="keep-words">
+            {network.device && networkState(network.device?.state)}
+          </Text>
+          {network.settings &&
+            <WifiNetworkMenu settings={network.settings} device={network.device} onConnect={onSelect} />}
+        </Split>
+      </Flex>
+      {((isSelected && (network && (!network.settings || network.error))) || network.needAuth) &&
```
Comment (imobachgs) [inline web/src/components/network/WifiNetworkListItemPage.jsx]: np: not to fix now, but this condition looks tricky to me.

File: web/src/components/network/WifiNetworksListPage.jsx [hunk 12]
```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (c) [2022] SUSE LLC
```
Comment (imobachgs) [inline web/src/components/network/WifiNetworksListPage.jsx:2]: Wrong year

File: web/src/components/network/WifiSelectorPage.jsx [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+            <WifiNetworksListPage
+              selected={selected}
+              onSelectionChange={selectNetwork}
+              networks={networksFromValues(networks)}
+              hiddenNetwork={baseHiddenNetwork}
+              activeNetwork={activeNetwork}
+              showHiddenForm={showHiddenForm}
+              availableNetworks={networks}
+              forceUpdateNetworksCallback={reloadNetworks}
+            // onSelectionCallback={(network) => {
```
Comment (imobachgs) [inline web/src/components/network/WifiSelectorPage.jsx:142]: Commented code.

File: web/src/components/network/WifiSelectorPage.jsx [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+const networksFromValues = (networks) => Object.values(networks).flat();
+const baseHiddenNetwork = { ssid: undefined, hidden: true };
+
+// FIXME: use a reducer
+
+function WifiSelectorPage() {
+  const { network: client } = useInstallerClient();
+  const { connections: initialConnections, devices: initialDevices, accessPoints, networks: initialNetworks } = useLoaderData();
+  const [data, saveData] = useLocalStorage("agama-network", { selectedWifi: null });
+  const [selected, setSelected] = useState(data.selectedWifi);
```
Comment (imobachgs) [inline web/src/components/network/WifiSelectorPage.jsx:42]: JFTR, we agreed to re-evaluate how to keep the state in the future (especially given that we are considering a state management library).

File: web/src/components/network/WifiSelectorPage.jsx [hunk 15]
```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) [2022] SUSE LLC
```
Comment (imobachgs) [inline web/src/components/network/WifiSelectorPage.jsx:2]: Wrong year.

File: web/src/components/network/WifiNetworkListItemPage.jsx (hunk 4 above)
Comment (teclator) [reply web/src/components/network/WifiNetworkListItemPage.jsx]: Removed file

File: web/src/components/network/NetworkPage.jsx (hunk 1 above)
Comment (teclator) [reply web/src/components/network/NetworkPage.jsx:55]: it was updateNetworks, as the method was get from WifiSelector and updateDevicesConnections was like too much.. was not very inspired at the moment of writing it :P

Comment (imobachgs) [reply web/src/components/network/NetworkPage.jsx:55]: As discussed offline, it is wrong because we are not checking the `updateState` value.

**Review by imobachgs — APPROVED** [review]

**Pull request merged**